eng/pipelines/templates/steps/test-conda.yml

```diff
@@ -0,0 +1,143 @@
+      versionSpec: $(PythonVersion)
+
+  - pwsh: |
+      # due to faulty deployed scripts/how the path gets manipulated by conda actions on 
+      # ubuntu and mac, we can't rely on bin/scripts being referenced correctly. see
+      # https://github.com/MicrosoftDocs/azure-devops-docs/issues/3812
+      $activateMethod = "source $($env:CONDA)/bin/activate"
+
+      # pypy3 is not a true python executable. in conda-land, we need to call it using pypy3, NOT python
+
+      # on windows, we need to add "--user" as otherwise pip won't successfully install/uninstall due to 
+      # how windows holds reservation on pip.exe. this is unnecessary on ubuntu/mac.
+      $requirementSuffix = ""
+
+      # we always want to prepend the path with conda bin   
+      Write-Host "##vso[task.prependpath]]$($env:CONDA)/bin"
+
+      if ($IsWindows) {
+        # powershell does not have an equivalent of call/source, which is necessary when
+        # using conda in azure devops. Note that we use `activate` natively here, as
+        # a later path prepend of the /scripts directory actually works.
+        $activateMethod = "call activate" 
+        $requirementSuffix = " --user"
+
+        # on windows only, need to prepend with the scripts directory as well
+        Write-Host "##vso[task.prependpath]$($env:CONDA)/Scripts"
+      }
+
+      if("$(PythonVersion)" -eq "pypy3"){
+        Write-Host "##vso[task.setvariable variable=PyVersion]-c conda-forge pypy3.7 pip"
+      }
+      else {
+        Write-Host "##vso[task.setvariable variable=PyVersion]python=$(PythonVersion)"
+      }
+      
+      # we will use these variables extensively later
+      Write-Host "##vso[task.setvariable variable=activate.method]$activateMethod"
+      Write-Host "##vso[task.setvariable variable=requirement.suffix]$requirementSuffix"
+    displayName: 'Evaluate OS Specific PATH and Parameters'
+
+  - ${{ each artifact in parameters.CondaArtifacts }}:
+    # due to the fact that `pypy3` and `conda-build` conda packages are INCOMPATIBLE, we have to create
+    # a separate env to install `conda-build` and use that to `conda index` the local file channel
+    - script: |
+        echo "conda create --name ${{ artifact.name }} $(PyVersion) --yes"
+        conda create --name ${{ artifact.name }} $(PyVersion) --yes
+
+        echo "conda create --name index-env --yes"
+        conda create --name index-env --yes
+
+        echo "conda install --name index-env --yes --quiet conda-build"
+        conda install --name index-env --yes --quiet conda-build
+
+        echo "$(activate.method) index-env"
+        $(activate.method) index-env
+
+        echo "conda index $(Build.ArtifactStagingDirectory)/${{ artifact.name }}"
+        conda index $(Build.ArtifactStagingDirectory)/${{ artifact.name }}
+      displayName: 'Prepare Conda Environment for Testing ${{ artifact.name }}, Index the Target Local Artifact'
+
+    - script: |
+        echo "$(activate.method) ${{ artifact.name }}"
```

Comment:
Given this is a script block instead of pwsh it will run on different shells on different OS's and echo is one of the commands that works different on linux vs windows.  There is a little more info on why at https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/2/Azure-DevOps-Pipeline-Guidance?anchor=use-full-cmdlet-names-for-better-scripts.

Reply:
What does it do differently? It's doing what I expect it to AFAICT.

[Example](https://dev.azure.com/azure-sdk/public/_build/results?buildId=875540&view=logs&j=7fdb0983-a0fb-582a-6142-53ff13da6d21&t=2d4deb93-1e3c-5a2c-90ed-c2018537416f)

Reply:
An extra `"` around it isn't the end of the world IMO.

Reply:
I agree the extra quote isn't the end of the world but that can start to fall apart when you are doing things like writing setvariable or other things that react different depending on the quoting around it. I know it bit me which is why I wrote about it in the wiki.

Reply:
Ok. I'm only using this to output the command that's being invoked for visibility. Thanks for a heads up on the setvariable stuff especially. Gonna look for any instances of that.